transaction/std/src/memo/gift_code_funding.rs

```diff
@@ -0,0 +1,310 @@
+// Copyright (c) 2018-2022 The MobileCoin Foundation
+
+//! Object for 0x0201 Gift Code Funding memo type
+//!
+//! This was proposed for standardization in mobilecoinfoundation/mcips/pull/32
+
+use crate::{impl_memo_type_conversions, RegisteredMemoType};
+use mc_crypto_hashes::{Blake2b512, Digest};
+use mc_crypto_keys::RistrettoPublic;
+use mc_transaction_core::MemoError;
+use std::{convert::TryInto, str};
+
+/// Mobilecoin account owners can create a special TxOut called a "gift code".
+/// This TxOut is sent to a special subaddress at index u64::MAX - 2 and the
+/// TxOut private key is sent to the intended recipient. This allows people who
+/// don't yet have a Mobilecoin account to receive Mobilecoin. When the sender
+/// makes the initial TxOut to the gift code subaddress, this memo will be
+/// written to the subaddress reserved for change TxOuts indicating that a gift
+/// code was funded. It includes the first 4 bytes of the hash of the TxOut to
+/// indicate which TxOut the gift code is at and 60 bytes representing a null
+/// terminated utf-8 string
+#[derive(Debug, Clone, Eq, PartialEq, Ord, PartialOrd)]
+pub struct GiftCodeFundingMemo {
+    /// The data representing the gift code memo
+    memo_data: [u8; Self::MEMO_DATA_LEN],
+}
+
+impl RegisteredMemoType for GiftCodeFundingMemo {
+    const MEMO_TYPE_BYTES: [u8; 2] = [0x02, 0x01];
+}
+
+impl GiftCodeFundingMemo {
+    /// Create a new gift funding code memo
+    pub fn new(tx_out_public_key: &RistrettoPublic, note: &str) -> Result<Self, MemoError> {
+        // Check if note is of valid length and initialize memo data
+        if note.len() > Self::NOTE_DATA_LEN {
+            return Err(MemoError::BadLength(note.len()));
+        }
+        let mut memo_data = [0u8; Self::MEMO_DATA_LEN];
+        // Compute TxOut hash and store it into the memo data
+        memo_data[0..Self::HASH_DATA_LEN]
+            .copy_from_slice(&tx_out_public_key_short_hash(tx_out_public_key));
+
+        // Put note into memo
+        let offset = Self::HASH_DATA_LEN;
+        memo_data[offset..(offset + note.len())].copy_from_slice(note.as_bytes());
+
+        Ok(Self { memo_data })
+    }
+
+    /// The Length of the TxOut hash
+    pub const HASH_DATA_LEN: usize = 4;
+
+    /// The length of the custom memo data
+    pub const MEMO_DATA_LEN: usize = 64;
+
+    /// Length of the utf-8 note
+    pub const NOTE_DATA_LEN: usize = 60;
+
+    /// Get the memo data
+    pub fn memo_data(&self) -> &[u8; Self::MEMO_DATA_LEN] {
+        &self.memo_data
+    }
+
+    /// Check if a given public key matches
+    pub fn public_key_matches(&self, tx_out_public_key: &RistrettoPublic) -> bool {
+        tx_out_public_key_short_hash(tx_out_public_key) == self.memo_data[0..Self::HASH_DATA_LEN]
+    }
+
+    /// Get funding note from memo
+    pub fn funding_note(&self) -> Result<&str, MemoError> {
+        let note = str::from_utf8(&self.memo_data[Self::HASH_DATA_LEN..])?;
```

Comment:
I think this is not the best way to implement this, because it's going to stick a bunch of null characters in the resulting string:

https://users.rust-lang.org/t/should-string-from-utf8-remove-trailing-null/52680

it would be better to do something where we find the index of the first null, if any, then change this to be like

```
let note = str::from_utf8(&self.memo_data[Self::HASH_DATA_LEN..first_null_index])?;
```

Reply:
although maybe i'm confused, because your unit test `test_gift_code_funding_memo_data_outputs_match()` seems to be passing...

Reply:
oh i see, you are doing the split_once thing after converting to utf-8

i think it might be marginally better to find the null before converting to utf-8, because it's conceivable that the stuff after the first null is not valid utf-8. but it's probably fine as is, i guess we'll always zero the unused bytes.

Reply:
Actually yeah, that IS the better way, I'm going to refactor to that